### ISSUE TITLE
some questions; info for reading

### ISSUE THREAD

**Issue by ian204** [lululxvi/deepxde#307]
Hello Lu,
Let me first express my thanks to you for bringing to us this package which simplifies the application of NN to PDEs.
My interests are on solving time-dependent NS equations, with various BC and IC. I have read your guide and FAQ list but I need some extra readings for understanding better DeepXDE package. If you, or anyone else in this group, can answer/comment my following questions I would appreciate very much.
Thank you.
Ian

Q1: Where I can find proper description of the syntax for using  "lambda x, or lambda x, y or lambda_" ? This would help to better understand the implementation of soft/hard BC/IC.
Q2: How can I use my optimization function to minimize the loss function?
Q3: How can I access the gradient of loss function wrt to all variables (weights and biases) or with respect to a subset of them?
Q4: How can I access the Hessian of the loss function wrt to a subset of variables (weights and biases)?
Q5: Does DeepXDE have the option to move the geometry?
Q6: Is there any explicit example for time-dependent NS equations (2D/3D)?

**Comment by lululxvi** [comment]
- Q1; This is standard Python grammar. You should find lots of tutorials online.
- Q2:I have no experience with this. You need to figure out how to write your own optimizer instead of those in Tensorflow.
- Q3 &Q4: I also don't have experience with this. There should a way in tensorflow. Or actually you can implement the network from scratch (instead of using something like tf.layers), which is very easy to do for fully-connected networks, and then you have the full control, and can directly use `tf.gradients` to get the gradient. Apply twice to get Hessian.
- Q5: what do you mean by "move the geometry"?
- Q6: I currently don't have NS equation example, but the implementation of NS is quite similar to other problems.

**Comment by ian204** [comment]
Thank you for the comments, Lu.
I have added some further explanation as it seems that I didn't describe properly my question in my original message.

Ian 

For Q2: 
Definitely, it is me who will write my own optimization function, no issue here. The question is how to call this new function instead of the ones tensorflow uses.
For Q5:
I mean how one could modify the domain. Think of NS equation in a moving domain ...
For Q6:
Actually, it seems the the code with Beltrami flow is a NS code, just with specific data

**Comment by lululxvi** [comment]
Here are some details of the optimizer: https://www.tensorflow.org/api_docs/python/tf/compat/v1/train/Optimizer#apply_gradients You can use `compute_gradients` to get the gradients, and I think you need to implement your own `apply_gradients`. So basically you need to write a subclass of `tf.compat.v1.train.Optimizer`

For moving domain, you can use a sufficiently large domain that covers your true domain. Then instead of asking DeepXDE to sample the training points, you can sample the points `X` by yourself, and then define `PDE(..., anchors=X)`. Or you can implement a subclass of [`Geometry`](https://deepxde.readthedocs.io/en/latest/modules/deepxde.geometry.html#deepxde.geometry.geometry.Geometry)

**Comment by ian204** [comment]
Thank you Lu. It gives me some ideas to work on.
